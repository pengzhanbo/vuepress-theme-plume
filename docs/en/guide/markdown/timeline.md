---
title: Timeline
icon: mdi:timeline-text-outline
createTime: 2025/03/24 21:19:12
permalink: /en/guide/markdown/timeline/
badge:
  text: 1.0.0-rc.137 +
  type: tip
---

## Overview

In markdown, using the `::: timeline` container with markdown unordered list syntax achieves a ==timeline== rendering effect.

- Supports ==horizontal== and ==vertical== directions
- Vertical direction supports **left alignment**, **right alignment**, and **both ends alignment**
- Supports **icons** and **line styles**
- Supports preset **types** to set **colors** and custom colors

## Enablement

This feature is not enabled by default. You need to enable it in the `theme` configuration.

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      timeline: true, // [!code ++]
    }
  })
})
```

## Usage

In the `::: timeline` container, use markdown unordered list syntax, with each list item being a point on the timeline.

```md{1,9} title="timeline.md"
::: timeline Configuration
- Title
  Configuration

  Body content

- Title
  Configuration

  Body content
:::
```

For each list item:

- From **the first line** to **the first blank line**, it is the **title**, with the line following the title used to **configure** the item's behavior
- **After the first blank line**: body content

:::important Please note to add the correct indentation
:::

_A simple example:_

_Input:_

```md
::: timeline
- Node one
  time=2025-03-20 type=success

  Body content

- Node two
  time=2025-02-21 type=warning

  Body content

- Node three
  time=2025-01-22 type=danger

  Body content
:::
```

_Output:_

::: timeline

- Node one
  time=2025-03-20 type=success

  Body content

- Node two
  time=2025-02-21 type=warning

  Body content

- Node three
  time=2025-01-22 type=danger

  Body content
  :::

::: important The timeline defaults to vertical direction
:::

## Configuration

The **timeline** supports flexible configuration options, divided into two main parts:

- **Container configuration**: Configurations on the `::: timeline` container, following `::: timeline` like `::: timeline horizontal` for horizontal timeline rendering.
- **List item configuration**: Configurations for each list item, following the title line like:
  ```md
  ::: timeline
  - Title                          <!--Title line-->
    Also title                      <!--Title line-->
    time=2025-03-20 type=success  <!--Configuration line, optional-->
                                  <!--Blank line, required if there's body content-->
    Body content
  :::
  ```

### Container Configuration

#### horizontal

- **Type**: `boolean`
- **Default**: `false`

Renders the timeline horizontally.

#### card

- **Type**: `boolean`
- **Default**: `false`

Renders each timeline node as a card style (can be overridden in list item configuration).

#### placement

- **Type**: `'left' | 'right' | 'between'`
- **Default**: `'left'`

Alignment of timeline nodes ==only effective in vertical direction=={.warning}.

- `left`: Aligns timeline to the left
- `right`: Aligns timeline to the right
- `between`: Aligns timeline to both ends (define position via `placement` in list item configuration, default is `left`)

#### line

- **Type**: `'solid' | 'dashed' | 'dotted'`
- **Default**: `'solid'`

Line style (can be overridden in list item configuration).

### List Item Configuration

#### time

- **Type**: `string`
- **Default**: `''`

Time point, can be any string like `2025-03-20` or `Q1`.

#### type

- **Type**: `'info' | 'tip' | 'success' | 'warning' | 'danger' | 'caution' | 'important'`
- **Default**: `'info'`

Type of timeline node.

#### card

- **Type**: `boolean`
- **Default**: `false` (inherits from container configuration)

Renders the current node as a card style.

#### line

- **Type**: `'solid' | 'dashed' | 'dotted'`
- **Default**: `'solid'` (inherits from container configuration)

Line style.

#### icon

- **Type**: `string`
- **Default**: `''`

Icon for the timeline node, supports all [iconify](https://icon-sets.iconify.design/) icons.

#### placement

- **Type**: `'left' | 'right'`
- **Default**: `'left'`

Position of the node when container is set to `between`.

- `left`: On the left side of the timeline
- `right`: On the right side of the timeline

#### color

- **Type**: `string`
- **Default**: `''`

Color for the timeline node line, can be any valid color value.

## Examples

### Horizontal Direction

Add `horizontal` after `:::timeline` to render the timeline horizontally.

Input:

```md /horizontal/
::: timeline horizontal
- Node one
  time=2025-03-20

  Body content

- Node two
  time=2025-04-20 type=success

  Body content

- Node three
  time=2025-01-22 type=danger

  Body content

- Node four
  time=2025-01-22 type=important

  Body content
:::
```

Output:

::: timeline horizontal

- Node one
  time=2025-03-20

  Body content

- Node two
  time=2025-04-20 type=success

  Body content

- Node three
  time=2025-01-22 type=danger

  Body content

- Node four
  time=2025-01-22 type=important

  Body content
  :::

### Right Alignment

Add `placement="right"` after `:::timeline` to render the timeline right-aligned.

Input:

```md /placement="right"/
::: timeline placement="right"
- Node one
  time=2025-03-20

  Body content

- Node two
  time=2025-04-20 type=success

  Body content

- Node three
  time=2025-01-22 type=danger

  Body content

- Node four
  time=2025-01-22 type=important

  Body content
:::
```

Output:

::: timeline placement="right"

- Node one
  time=2025-03-20

  Body content

- Node two
  time=2025-04-20 type=success

  Body content

- Node three
  time=2025-01-22 type=danger

  Body content

- Node four
  time=2025-01-22 type=important

  Body content
  :::

### Both Ends Alignment

Add `placement="between"` after `:::timeline` to render the timeline aligned to both ends.

List items default to the left side of the timeline, use `placement="right"` in list item configuration to set the position to the right.

Input:

```md /placement="between"/ /placement=right/
::: timeline placement="between"
- Node one
  time=2025-03-20 placement=right

  Body content

- Node two
  time=2025-04-20 type=success

  Body content

- Node three
  time=2025-01-22 type=danger placement=right

  Body content

- Node four
  time=2025-01-22 type=important

  Body content
:::
```

Output:

::: timeline placement="between"

- Node one
  time=2025-03-20 placement=right

  Body content

- Node two
  time=2025-04-20 type=success

  Body content

- Node three
  time=2025-01-22 type=danger placement=right

  Body content

- Node four
  time=2025-01-22 type=important

  Body content
  :::

### Node Types

Add `type=node type` in list item configuration to set the node type.

Input:

```md /type=success/ /type=warning/ /type=danger/ /type=important/
::: timeline
- Node one
  time=2025-03-20 type=success

  Body content

- Node two
  time=2025-04-20 type=warning

  Body content

- Node three
  time=2025-01-22 type=danger

  Body content

- Node four
  time=2025-01-22 type=important

  Body content
:::
```

Output:

::: timeline

- Node one
  time=2025-03-20 type=success

  Body content

- Node two
  time=2025-04-20 type=warning

  Body content

- Node three
  time=2025-01-22 type=danger

  Body content

- Node four
  time=2025-01-22 type=important

  Body content
  :::

### Line Styles

- Add `line=line style` in container configuration to set the default line style for all nodes.
- Add `line=line style` in list item configuration to set the line style for individual nodes.

Input:

```md /line="dotted"/ /line=solid/ /line=dashed/
::: timeline line="dotted"
- Node one
  time=2025-03-20

  Body content

- Node two
  time=2025-04-20 type=success

  Body content

- Node three
  time=2025-01-22 type=danger line=dashed

  Body content

- Node four
  time=2025-01-22 type=important line=solid

  Body content
:::
```

Output:

::: timeline line="dotted"

- Node one
  time=2025-03-20

  Body content

- Node two
  time=2025-04-20 type=success

  Body content

- Node three
  time=2025-01-22 type=danger line=dashed

  Body content

- Node four
  time=2025-01-22 type=important line=solid

  Body content
  :::

### Nodes with Icons

Add `icon=icon name` in list item configuration to add icons to nodes.

Icon names support [iconify](https://icon-sets.iconify.design/) icon names.

Input:

```md /icon=mdi:balloon/ /icon=mdi:bookmark/
::: timeline placement="between"
- Node one
  time=2025-03-20 placement=right icon=mdi:balloon

  Body content

- Node two
  time=2025-04-20 type=success icon=mdi:bookmark

  Body content

- Node three
  time=2025-01-22 type=danger placement=right icon=mdi:bullhorn-variant-outline

  Body content

- Node four
  time=2025-01-22 type=important card=true icon="mdi:cake-variant-outline"

  Body content
:::
```

Output:

::: timeline placement="between"

- Node one
  time=2025-03-20 placement=right icon=mdi:balloon

  Body content

- Node two
  time=2025-04-20 type=success icon=mdi:bookmark

  Body content

- Node three
  time=2025-01-22 type=danger placement=right icon=mdi:bullhorn-variant-outline

  Body content

- Node four
  time=2025-01-22 type=important card=true icon="mdi:cake-variant-outline"

  Body content
  :::

### Card Nodes

Card nodes can be flexibly controlled:

- Add `card` in container configuration to make each list item a card node.
- Add `card=true` in list item configuration to set the node as a card node.
- Add `card=false` in list item configuration to set the node as a non-card node.

The style of card nodes is affected by the `type` configuration.

::: tip Adding `card=true` / `card=false` in list item configuration overrides the container's `card` configuration
:::

Input:

```md{1} /card=false/
::: timeline card
- Node one
  time=2025-03-20

  Body content

- Node two
  time=2025-04-20 type=success card=false

  Body content

- Node three
  time=2025-01-22 type=danger

  Body content

- Node four
  time=2025-01-22 type=important

  Body content
:::
```

Output:

::: timeline card

- Node one
  time=2025-03-20

  Body content

- Node two
  time=2025-04-20 type=success card=false

  Body content

- Node three
  time=2025-01-22 type=danger

  Body content

- Node four
  time=2025-01-22 type=important

  Body content
  :::

## Custom Node Types

Timeline node types are controlled via CSS Variables. The theme provides the following CSS variables:

```css
:root {
  --vp-timeline-c-line: var(--vp-c-border);   /* Line color */
  --vp-timeline-c-point: var(--vp-c-border);  /* Point color */
  --vp-timeline-c-title: var(--vp-c-text-1);  /* Title text color */
  --vp-timeline-c-text: var(--vp-c-text-1);   /* Body text color */
  --vp-timeline-c-time: var(--vp-c-text-3);   /* Time text color */
  --vp-timeline-c-icon: var(--vp-c-bg);       /* Icon color */
  --vp-timeline-bg-card: var(--vp-c-bg-soft); /* Card node background color */
}
```

For example, the theme's built-in node type `tip`:

```css /.tip/
.vp-timeline-item.tip {
  --vp-timeline-c-line: var(--vp-c-tip-1);
  --vp-timeline-c-point: var(--vp-c-tip-1);
  --vp-timeline-bg-card: var(--vp-c-tip-soft);
}
```

You can override built-in types or add new ones in [custom styles](../custom/style.md).

Example:

```css title=".vuepress/styles/index.css"
.vp-timeline-item.your-type {
  --vp-timeline-c-line: #3cf;
  --vp-timeline-c-point: #3cf;
  --vp-timeline-bg-card: rgba(60, 252, 255, 0.314);
}
```

```md /type=your-type/
::: timeline
- Node one
  time=2025-03-20

  Body content

- Node two
  time=2025-04-20 type=your-type card=true

  Body content

- Node three
  time=2025-01-22 type=danger

  Body content
:::
```

::: timeline

- Node one
  time=2025-03-20

  Body content

- Node two
  time=2025-04-20 type=your-type card=true

  Body content

- Node three
  time=2025-01-22 type=danger

  Body content
  :::

<style>
.vp-timeline-item.your-type {
  --vp-timeline-c-line: #3cf;
  --vp-timeline-c-point: #3cf;
  --vp-timeline-bg-card: rgba(60, 252, 255, 0.314);
}
</style>
