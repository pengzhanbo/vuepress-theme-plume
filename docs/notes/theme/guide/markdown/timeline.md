---
title: 时间线
icon: mdi:timeline-text-outline
createTime: 2025/03/20 18:05:29
permalink: /guide/markdown/timeline/
badge:
  text: 1.0.0-rc.137 +
  type: tip
---

## 概述

在 markdown 中，使用 `::: timeline` 容器，包含 markdown 无序列表语法，即可实现 ==时间线== 的 渲染效果。

- 支持 ==水平方向== 和 ==垂直方向==
- 垂直方向支持 __左对齐__，__右对齐__ 和 __两端对齐__
- 支持 __图标__ 和 __线条样式__
- 支持 通过预设 __类型__ 设置 __颜色__，支持自定义颜色

## 启用

该功能默认不启用，你需要在 `theme` 配置中启用。

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      timeline: true, // [!code ++]
    }
  })
})
```

## 使用

在 `::: timeline` 容器中，使用 markdown 无序列表语法，列表的每一个项即 时间线上的每一个点。

```md{1,9} title="timeline.md"
::: timeline 配置
- 标题
  配置

  正文内容

- 标题
  配置

  正文内容
:::
```

对于列表的每一个项：

- 从 __首行开始__ 到 __首个空行__，均为 __标题__ ，在标题后紧跟随的一行，用于 __配置__ 当前项的行为

- __首个空行之后__: 正文内容

:::important 请注意添加正确的缩进
:::

__一个简单的例子：__

__输入：__

```md
::: timeline
- 节点一
  time=2025-03-20 type=success

  正文内容

- 节点二
  time=2025-02-21 type=warning

  正文内容

- 节点三
  time=2025-01-22 type=danger

  正文内容
:::
```

__输出：__

::: timeline

- 节点一
  time=2025-03-20 type=success

  正文内容

- 节点二
  time=2025-02-21 type=warning

  正文内容

- 节点三
  time=2025-01-22 type=danger

  正文内容
:::

::: important 时间线默认为垂直方向
:::

## 配置

__时间线__ 支持非常灵活且灵活的配置项，配置主要分为两个部分：

- __容器配置__： 在 `::: timeline` 容器上的配置，配置项跟随在 `::: timeline` 之后，如：

  `::: timeline horizontal` 表示 渲染为 水平方向的时间线。

- __列表项配置__： 列表的每一个项的配置，紧跟随在标题之后的一行，如：

  ```md
  ::: timeline
  - 标题                          <!--标题行-->
    也是标题                      <!--标题行-->
    time=2025-03-20 type=success  <!--配置跟随在最后的标题行之后的单独一行，可选-->
                                  <!--空行，有正文时必须-->
    正文内容
  :::
  ```

### 容器配置

#### horizontal

- __类型:__ `boolean`
- __默认值:__ `false`

渲染为 水平方向的时间线。

#### card

- __类型:__ `boolean`
- __默认值:__ `false`

每个时间节点默认渲染为卡片样式（可在列表项配置中覆盖）。

#### placement

- __类型:__ `'left' | 'right' | 'between'`
- __默认值:__ `'left'`

时间节点的对齐方式。==仅在垂直方向时生效=={.warning}

- `left` : 时间轴左侧对齐
- `right` : 时间轴右侧对齐
- `between` : 时间轴两端对齐 (通过列表项配置中的 `placement` 定义位置，默认为 `left`)

#### line

- __类型:__ `'solid' | 'dashed' | 'dotted'`
- __默认值:__ `'solid'`

线条样式（可在列表项配置中覆盖）

### 列表项配置

#### time

- __类型:__ `string`
- __默认值:__ `''`

时间点，可以是任何字符串，比如 `2025-03-20`， `Q1` 等。

#### type

- __类型:__ `'info' | 'tip' | 'success' | 'warning' | 'danger' | 'caution' | 'important'`
- __默认值:__ `'info'`

时间节点的类型。

#### card

- __类型:__ `boolean`
- __默认值:__ `false` 从 容器配置 `card` 中继承

当前 时间节点渲染为卡片样式。

#### line

- __类型:__ `'solid' | 'dashed' | 'dotted'`
- __默认值:__ `'solid'` 从 容器配置 `line` 中继承

线条样式

#### icon

- __类型:__ `string`
- __默认值:__ `''`

时间节点的图标，支持所有的 [iconify](https://icon-sets.iconify.design/) 图标。

#### placement

- __类型:__ `'left' | 'right'`
- __默认值:__ `'left'`

当 容器配置为 `between` 时，定义当前时间节点的位置。

- `left` : 在时间轴左侧
- `right` : 在时间轴右侧

#### color

- __类型:__ `string`
- __默认值:__ `''`

时间节点线条颜色，可以是任何有效的颜色值。

## 示例

### 水平方向

在 `:::timeline` 后跟随声明 `horizontal` , 即可将时间线渲染为 水平方向。

__输入：__

```md /horizontal/
::: timeline horizontal
- 节点一
  time=2025-03-20

  正文内容

- 节点二
  time=2025-04-20 type=success

  正文内容

- 节点三
  time=2025-01-22 type=danger

  正文内容

- 节点四
  time=2025-01-22 type=important

  正文内容
:::
```

__输出：__

::: timeline horizontal

- 节点一
  time=2025-03-20

  正文内容

- 节点二
  time=2025-04-20 type=success

  正文内容

- 节点三
  time=2025-01-22 type=danger

  正文内容

- 节点四
  time=2025-01-22 type=important

  正文内容
:::

### 右对齐

在 `:::timeline` 后跟随声明 `placement="right"` , 即可将时间线渲染为 右对齐。

__输入：__

```md /placement="right"/
::: timeline placement="right"
- 节点一
  time=2025-03-20

  正文内容

- 节点二
  time=2025-04-20 type=success

  正文内容

- 节点三
  time=2025-01-22 type=danger

  正文内容

- 节点四
  time=2025-01-22 type=important

  正文内容
:::
```

__输出：__

::: timeline placement="right"

- 节点一
  time=2025-03-20

  正文内容

- 节点二
  time=2025-04-20 type=success

  正文内容

- 节点三
  time=2025-01-22 type=danger

  正文内容

- 节点四
  time=2025-01-22 type=important

  正文内容
:::

### 两端对齐

在 `:::timeline` 后跟随声明 `placement="between"` , 即可将时间线渲染为 两端对齐。

列表项默认位于时间线的左侧，可以通过 `placement="right"` 为列表项设置右侧位置。

__输入：__

```md /placement="between"/ /placement=right/
::: timeline placement="between"
- 节点一
  time=2025-03-20 placement=right

  正文内容

- 节点二
  time=2025-04-20 type=success

  正文内容

- 节点三
  time=2025-01-22 type=danger placement=right

  正文内容

- 节点四
  time=2025-01-22 type=important

  正文内容
:::
```

__输出：__

::: timeline placement="between"

- 节点一
  time=2025-03-20 placement=right

  正文内容

- 节点二
  time=2025-04-20 type=success

  正文内容

- 节点三
  time=2025-01-22 type=danger placement=right

  正文内容

- 节点四
  time=2025-01-22 type=important

  正文内容
:::

### 节点类型

在列表项配置中，添加 `type=节点类型` 可以为当前节点设置节点类型。

__输入：__

```md /type=success/ /type=warning/ /type=danger/ /type=important/
::: timeline
- 节点一
  time=2025-03-20 type=success

  正文内容

- 节点二
  time=2025-04-20 type=warning

  正文内容

- 节点三
  time=2025-01-22 type=danger

  正文内容

- 节点四
  time=2025-01-22 type=important

  正文内容
:::
```

__输出：__

::: timeline

- 节点一
  time=2025-03-20 type=success

  正文内容

- 节点二
  time=2025-04-20 type=warning

  正文内容

- 节点三
  time=2025-01-22 type=danger

  正文内容

- 节点四
  time=2025-01-22 type=important

  正文内容
:::

### 线条风格

- 在容器配置中添加 `line=线条风格` 可以为所有节点设置默认线条风格。
- 在列表项配置中，添加 `line=线条风格` 可以为节点设置线条风格。

__输入：__

```md /line="dotted"/ /line=solid/ /line=dashed/
::: timeline line="dotted"
- 节点一
  time=2025-03-20

  正文内容

- 节点二
  time=2025-04-20 type=success

  正文内容

- 节点三
  time=2025-01-22 type=danger line=dashed

  正文内容

- 节点四
  time=2025-01-22 type=important line=solid

  正文内容
:::
```

__输出：__

::: timeline line="dotted"

- 节点一
  time=2025-03-20

  正文内容

- 节点二
  time=2025-04-20 type=success

  正文内容

- 节点三
  time=2025-01-22 type=danger line=dashed

  正文内容

- 节点四
  time=2025-01-22 type=important line=solid

  正文内容
:::

### 带图标的节点

在列表项配置中，添加 `icon=图标名称` 可以为节点添加图标。

图标名称支持 [iconify](https://icon-sets.iconify.design/) 的图标名称。

__输入：__

```md /icon=mdi:balloon/ /icon=mdi:bookmark/
::: timeline placement="between"
- 节点一
  time=2025-03-20 placement=right icon=mdi:balloon

  正文内容

- 节点二
  time=2025-04-20 type=success icon=mdi:bookmark

  正文内容

- 节点三
  time=2025-01-22 type=danger placement=right icon=mdi:bullhorn-variant-outline

  正文内容

- 节点四
  time=2025-01-22 type=important card=true icon="mdi:cake-variant-outline"

  正文内容
:::
```

__输出：__

::: timeline placement="between"

- 节点一
  time=2025-03-20 placement=right icon=mdi:balloon

  正文内容

- 节点二
  time=2025-04-20 type=success icon=mdi:bookmark

  正文内容

- 节点三
  time=2025-01-22 type=danger placement=right icon=mdi:bullhorn-variant-outline

  正文内容

- 节点四
  time=2025-01-22 type=important card=true icon="mdi:cake-variant-outline"

  正文内容
:::

### 卡片节点

卡片节点可以很灵活的进行控制：

- 在 容器配置中添加 `card` 即可使每个列表项都是卡片节点。
- 在列表项配置中，添加 `card=true` 即可为节点设置为卡片节点。
- 在列表项配置中，添加 `card=false` 即可为节点设置为非卡片节点。

卡片节点的样式会受到 `type` 配置的影响。

::: tip 在列表项配置中添加 `card=true` / `card=false` 可以覆盖容器节点的 `card` 配置
:::

__输入：__

```md{1} /card=false/
::: timeline card
- 节点一
  time=2025-03-20

  正文内容

- 节点二
  time=2025-04-20 type=success card=false

  正文内容

- 节点三
  time=2025-01-22 type=danger

  正文内容

- 节点四
  time=2025-01-22 type=important

  正文内容
:::
```

__输出：__

::: timeline card

- 节点一
  time=2025-03-20

  正文内容

- 节点二
  time=2025-04-20 type=success card=false

  正文内容

- 节点三
  time=2025-01-22 type=danger

  正文内容

- 节点四
  time=2025-01-22 type=important

  正文内容
:::

## 自定义节点类型

时间轴的节点类型是通过 CSS Variables 控制的，主题提供了以下的 CSS 变量：

```css
:root {
  --vp-timeline-c-line: var(--vp-c-border);   /* 线条颜色 */
  --vp-timeline-c-point: var(--vp-c-border);  /* 点颜色 */
  --vp-timeline-c-title: var(--vp-c-text-1);  /* 标题文本颜色 */
  --vp-timeline-c-text: var(--vp-c-text-1);   /* 正文文本颜色 */
  --vp-timeline-c-time: var(--vp-c-text-3);   /* 时间文本颜色 */
  --vp-timeline-c-icon: var(--vp-c-bg);       /* 图标颜色 */
  --vp-timeline-bg-card: var(--vp-c-bg-soft); /* 卡片节点的背景颜色 */
}
```

比如主题内置的节点类型 `tip`:

```css /.tip/
.vp-timeline-item.tip {
  --vp-timeline-c-line: var(--vp-c-tip-1);
  --vp-timeline-c-point: var(--vp-c-tip-1);
  --vp-timeline-bg-card: var(--vp-c-tip-soft);
}
```

可以在 [自定义样式](../custom/style.md) 中，覆盖内置的类型，或者添加新的类型。

__示例：__

```css title=".vuepress/styles/index.css"
.vp-timeline-item.your-type {
  --vp-timeline-c-line: #3cf;
  --vp-timeline-c-point: #3cf;
  --vp-timeline-bg-card: rgba(60, 252, 255, 0.314);
}
```

```md /type=your-type/
::: timeline
- 节点一
  time=2025-03-20

  正文内容

- 节点二
  time=2025-04-20 type=your-type card=true

  正文内容

- 节点三
  time=2025-01-22 type=danger

  正文内容
:::
```

::: timeline

- 节点一
  time=2025-03-20

  正文内容

- 节点二
  time=2025-04-20 type=your-type card=true

  正文内容

- 节点三
  time=2025-01-22 type=danger

  正文内容
:::

<style>
.vp-timeline-item.your-type {
  --vp-timeline-c-line: #3cf;
  --vp-timeline-c-point: #3cf;
  --vp-timeline-bg-card: rgba(60, 252, 255, 0.314);
}
</style>
