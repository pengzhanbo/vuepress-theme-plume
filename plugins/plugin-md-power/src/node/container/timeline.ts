/**
 * ::: timeline
 *
 * - title
 *   time="Q1" icon="ri:clockwise-line" line="dashed" type="warning" color="red"
 *
 *   content
 *
 * - title
 *   time="Q2" icon="ri:clockwise-line" line="dashed" type="warning" color="red"
 *
 *   content
 * :::
 */
import type Token from 'markdown-it/lib/token.mjs'
import type { Markdown } from 'vuepress/markdown'
import { isEmptyObject } from '@pengzhanbo/utils'
import { resolveAttrs } from '.././utils/resolveAttrs.js'
import { stringifyAttrs } from '../utils/stringifyAttrs.js'
import { createContainerPlugin } from './createContainer.js'

export interface TimelineAttrs {
  horizontal?: boolean
  card?: boolean
  placement?: string
  line?: string
}

export interface TimelineItemMeta {
  time?: string
  type?: string
  icon?: string
  color?: string
  line?: string
  card?: string
  placement?: string
}

const RE_KEY = /(\w+)=\s*/
const RE_SEARCH_KEY = /\s+\w+=\s*|$/
const RE_CLEAN_VALUE = /(?<quote>["'])(.*?)(\k<quote>)/

export function timelinePlugin(md: Markdown): void {
  createContainerPlugin(md, 'timeline', {
    before(info, tokens, index) {
      parseTimeline(tokens, index)

      const { attrs } = resolveAttrs<TimelineAttrs>(info)
      attrs.card ??= undefined
      return `<VPTimeline${stringifyAttrs(attrs, true)}>`
    },
    after: () => '</VPTimeline>',
  })

  md.renderer.rules.timeline_item_open = (tokens, idx) => {
    const token = tokens[idx]
    const attrs = token.meta as TimelineItemMeta
    attrs.card ??= undefined
    const icon = attrs.icon
    return `<VPTimelineItem${stringifyAttrs(attrs, true)}>${icon ? `<template #icon><VPIcon name="${icon}"/></template>` : ''}`
  }

  md.renderer.rules.timeline_item_close = () => '</VPTimelineItem>'
  md.renderer.rules.timeline_item_title_open = () => '<template #title>'
  md.renderer.rules.timeline_item_title_close = () => '</template>'
}

function parseTimeline(tokens: Token[], index: number) {
  const listStack: number[] = [] // 记录列表嵌套深度

  for (let i = index + 1; i < tokens.length; i++) {
    const token = tokens[i]
    if (token.type === 'container_timeline_close') {
      break
    }
    // 列表层级追踪
    if (token.type === 'bullet_list_open') {
      listStack.push(0) // 每个新列表初始层级为0
      if (listStack.length === 1)
        token.hidden = true
    }
    else if (token.type === 'bullet_list_close') {
      listStack.pop()
      if (listStack.length === 0)
        token.hidden = true
    }
    else if (token.type === 'list_item_open') {
      const currentLevel = listStack.length
      // 仅处理根级列表项（层级1）
      if (currentLevel === 1) {
        token.type = 'timeline_item_open'
        tokens[i + 1].type = 'timeline_item_title_open'
        tokens[i + 3].type = 'timeline_item_title_close'

        // - title
        //   attrs
        // 列表项 `-` 后面包括紧跟随的后续行均在 type=inline 的 token 中， 并作为 children
        const inlineToken = tokens[i + 2]
        // 找到最后一个 softbreak，最后一行作为 attrs 进行解析
        const softbreakIndex = inlineToken.children!.findLastIndex(
          token => token.type === 'softbreak',
        )
        if (softbreakIndex !== -1) {
          const lastToken = inlineToken.children![inlineToken.children!.length - 1]
          token.meta = extractTimelineAttributes(lastToken.content.trim())
          if (!isEmptyObject(token.meta)) {
            inlineToken.children = inlineToken.children!.slice(0, softbreakIndex)
          }
        }
        else {
          token.meta = {}
        }
      }
    }
    else if (token.type === 'list_item_close') {
      const currentLevel = listStack.length
      if (currentLevel === 1) {
        token.type = 'timeline_item_close'
      }
    }
  }
}

export function extractTimelineAttributes(rawText: string): TimelineItemMeta {
  const attrKeys = ['time', 'type', 'icon', 'line', 'color', 'card', 'placement'] as const
  const attrs: TimelineItemMeta = {}
  let buffer = rawText.trim()

  while (buffer.length) {
    // 匹配属性键 (支持大小写)
    const keyMatch = buffer.match(RE_KEY)
    if (!keyMatch) {
      break
    }

    // 提取可能的关键字
    const matchedKey = keyMatch[1].toLowerCase()
    if (!attrKeys.includes(matchedKey as any)) {
      break
    }
    const keyStart = keyMatch.index!

    // 跳过已匹配的 key:
    const keyEnd = keyStart + keyMatch[0].length
    buffer = buffer.slice(keyEnd)

    // 提取属性值 (到下一个属性或行尾)
    let valueEnd = buffer.search(RE_SEARCH_KEY)
    /* istanbul ignore if -- @preserve */
    if (valueEnd === -1)
      valueEnd = buffer.length
    const value = buffer.slice(0, valueEnd).trim()
    // 存储属性
    attrs[matchedKey as keyof TimelineItemMeta] = value.replace(RE_CLEAN_VALUE, '$2')

    // 跳过已处理的值
    buffer = buffer.slice(valueEnd)
  }

  return attrs
}
