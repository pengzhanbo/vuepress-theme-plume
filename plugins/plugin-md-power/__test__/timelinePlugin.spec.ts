import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { extractTimelineAttributes, timelinePlugin } from '../src/node/container/timeline.js'

describe('timeline > extractTimelineAttributes()', () => {
  it('should work', () => {
    const meta = extractTimelineAttributes('这time=Q1')
    expect(meta).toEqual({ time: 'Q1' })
  })

  it('should work with multi attrs', () => {
    const meta = extractTimelineAttributes('time=Q1 icon=ri:clockwise-line card=true placement=left line=dashed')
    expect(meta).toEqual({ time: 'Q1', icon: 'ri:clockwise-line', card: 'true', placement: 'left', line: 'dashed' })
  })

  it('should work with title include space', () => {
    const meta = extractTimelineAttributes('time=Q1 icon=ri:clockwise-line card=true placement=left line=dashed')

    expect(meta).toEqual({ time: 'Q1', icon: 'ri:clockwise-line', card: 'true', placement: 'left', line: 'dashed' })
  })

  it('should work with unknown attr', () => {
    const meta = extractTimelineAttributes('time=Q1 unknown=true card=true')
    expect(meta).toEqual({ time: 'Q1' })
  })
})

describe('timeline > timelinePlugin()', () => {
  const md = new MarkdownIt()
  timelinePlugin(md)

  it('should work', () => {
    const source = `\
::: timeline
- 这是标题

  这是内容

- 这是标题
  这也是标题

  这是内容
:::

::: timeline horizontal line="dashed" card
- 这是标题
  time=q1

  这是内容
  - 1
  - 2
  - 3
    - 1.1
    - 1.2

- 这是标题
  time=q2 color=red card=false

  这是内容
:::

::: timeline placement="right"
- 这是标题
  icon=xxx card=true type=warning

  这是内容

- 这是标题
  type=danger line=dotted

  这是内容
:::

::: timeline placement="between"
- 这是标题
  card=true placement=right

  这是内容

- 这是标题
  card=true placement=left

  这是内容

- 这是标题

  这是内容
:::
`
    expect(md.render(source)).toMatchSnapshot()
  })
})
