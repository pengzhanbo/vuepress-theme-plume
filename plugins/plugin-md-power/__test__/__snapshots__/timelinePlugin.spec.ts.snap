// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`timeline > timelinePlugin() > should work 1`] = `
"<VPTimeline :card="undefined"><VPTimelineItem :card="undefined"><template #title>这是标题</template><p>这是内容</p>
</VPTimelineItem><VPTimelineItem :card="undefined"><template #title>这是标题
这也是标题</template><p>这是内容</p>
</VPTimelineItem></VPTimeline><VPTimeline horizontal line="dashed" card><VPTimelineItem time="q1" :card="undefined"><template #title>这是标题</template><p>这是内容</p>
<ul>
<li>1</li>
<li>2</li>
<li>3
<ul>
<li>1.1</li>
<li>1.2</li>
</ul>
</li>
</ul>
</VPTimelineItem><VPTimelineItem time="q2" color="red"><template #title>这是标题</template><p>这是内容</p>
</VPTimelineItem></VPTimeline><VPTimeline placement="right" :card="undefined"><VPTimelineItem icon="xxx" card type="warning"><template #icon><VPIcon name="xxx"/></template><template #title>这是标题</template><p>这是内容</p>
</VPTimelineItem><VPTimelineItem type="danger" line="dotted" :card="undefined"><template #title>这是标题</template><p>这是内容</p>
</VPTimelineItem></VPTimeline><VPTimeline placement="between" :card="undefined"><VPTimelineItem card placement="right"><template #title>这是标题</template><p>这是内容</p>
</VPTimelineItem><VPTimelineItem card placement="left"><template #title>这是标题</template><p>这是内容</p>
</VPTimelineItem><VPTimelineItem :card="undefined"><template #title>这是标题</template><p>这是内容</p>
</VPTimelineItem></VPTimeline>"
`;
